rts/game_MC/mc_rule_actor.cc

```diff
@@ -24,8 +24,176 @@ bool MCRuleActor::ActByState2(const GameEnv &env, const vector<int>& state, stri
 
     return true;
 }
+const Unit* MCRuleActor::GetTargetUnit(const GameEnv &env, const vector<vector<const Unit*> > &enemy_troops, const Player& player){
+    const vector<UnitType> &attack_order = {BASE, DEFENSE_TOWER, FACTORY, BARRACK, HANGAR, WORKSHOP};
+    for (const UnitType &target_type : attack_order) {
+        const auto& target_units = enemy_troops[target_type];
+        if (!target_units.empty()) {
+            for (const Unit *u : target_units) {
+                // cout << "potential unit " << u->GetUnitType() << " at " << u->GetPointF();
+                if (player.FilterWithFOW(*u)) {
+                    return u;
+                }
+                Loc loc = env.GetMap().GetLoc(u->GetPointF().ToCoord());
+                const Fog &f = player.GetFog(loc);
+                if (!f.seen_units().empty()) {
+                    return u;
+                }
+            }
+        }
+    }
+    return nullptr;
+}
+
+bool MCRuleActor::ActByState(const GameEnv &env, const vector<int>& state, string* /*state_string*/, AssignedCmds *assigned_cmds) {
+    const GameDef& gamedef = env.GetGameDef();
+    const auto& my_troops = _preload.MyTroops();
+    const auto& enemy_troops = _preload.EnemyTroops();
+    const auto& enemy_troops_in_range = _preload.EnemyTroopsInRange();
+    const auto& all_my_troops = _preload.AllMyTroops();
+    const auto& enemy_attacking_economy = _preload.EnemyAttackingEconomy();
+    const Player& player = env.GetPlayer(_player_id);
+    // cout << "entering act by state" << endl << flush;
+    if (state[STATE_GATHER]) {
+        for (const Unit *u : my_troops[WORKER]) {
+            if (IsIdle(*_receiver, *u)) {
+                // Gather!
+                float dist, closest_dist=1e10;
+                UnitId base_id, resource_id;
+                for (const Unit* resource : my_troops[RESOURCE]) {
+                    // check closest resource
+                    PointF resource_loc = resource->GetPointF();
+                    UnitId closest_base_id = env.FindClosestBase(_player_id, resource_loc, &dist);
+                    if (dist < closest_dist) {
+                        closest_dist = dist;
+                        base_id = closest_base_id;
+                        resource_id = resource->GetId();
+                    }
+                }
+                store_cmd(u, _G(base_id, resource_id), assigned_cmds);
+            }
+        }
+    }
+    //cout << "after gather" << endl << flush;
+    for (int build_state = (int)STATE_BUILD_WORKER; build_state < (int)STATE_BUILD_BASE; build_state++) {
```

Comment:
better to add a vector of states, rather than rely on the fact that the states will be contiguous